compiler/rustc_middle/src/ty/pattern.rs

```diff
@@ -27,17 +28,37 @@ impl<'tcx> fmt::Debug for PatternKind<'tcx> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match *self {
             PatternKind::Range { start, end, include_end } => {
-                if let Some(start) = start {
-                    write!(f, "{start}")?;
-                }
-                write!(f, "..")?;
-                if include_end {
-                    write!(f, "=")?;
-                }
-                if let Some(end) = end {
-                    write!(f, "{end}")?;
+                write!(f, "{start}")?;
+
+                match include_end {
+                    RangeEnd::Included => {
+                        if let Some(c) = end.try_to_value() {
+                            let end = c.valtree.unwrap_leaf();
+                            let size = end.size();
+                            let max = match c.ty.kind() {
+                                ty::Int(_) => Some(ty::ScalarInt::truncate_from_int(
+                                    size.signed_int_max(),
+                                    size,
+                                )),
+                                ty::Uint(_) => Some(ty::ScalarInt::truncate_from_uint(
+                                    size.unsigned_int_max(),
+                                    size,
+                                )),
+                                ty::Char => {
+                                    Some(ty::ScalarInt::truncate_from_uint(char::MAX, size))
+                                }
+                                _ => None,
+                            };
+                            if let Some((max, _)) = max
+                                && end == max
+                            {
+                                return write!(f, "..");
+                            }
+                        }
+                    }
+                    RangeEnd::Excluded => {}
                 }
-                Ok(())
+                write!(f, "{include_end}{end}")
             }
```

Comment:
Could have done this on master to avoid printing `1..=` for `1..` ranges, but it only got annoying once this PR made it `1..=u32::MAX` :D